gist.go

```diff
@@ -0,0 +1,86 @@
+package model
+
+// Typed constants
+type ChannelType string
+
+// CamelCase Constant Names
+const (
+	ChannelTypeOpen    ChannelType = "O"
+	ChannelTypePrivate ChannelType = "P"
+)
+
+type Channel struct {
+	// Member variables are not exported to ensure immutability.
+	id       string `json:"id"`
+	name     string `json:"name"`
+	createAt int64  `json:"create_at"`
+	// etc...
+
+	// Question: what do we do when there's a semantically meaningful
+	// "null" value for a member variable?
+	schemeID *string `json:"scheme_id"`
+
+	// Question: is this the best way of storing props and similar map/list data?
+	props map[string]interface{} `json:"props" db:"-"`
+}
+
+// Getters to access the member variables are provided. These *never* return pointers.
+// Correct capitalisation of initialisms
+func (channel *Channel) GetID() string {
+	return channel.id
+}
+
+// Returning `interface{}` feels gross. Can we do better than this?
+func (channel *Channel) GetProp(key string) interface{} {
+	return channel.props[key]
+}
+
+// "Patch" object holds changes to the object to make.
+type ChannelPatch struct {
+	Name *string `json:"name"`
+	// etc...
+
+	// Question: can we avoid pointers here and have some type like "NullableString"
+	// that indicates whether the field should be set explicitly instead?
+}
+
+// Similarly to what we do for the "PATCH" API endpoints currently, we modify model
+// objects by populating a "patch" object then calling a "ApplyPatch" method.
+func (channel *Channel) ApplyPatch(patch ChannelPatch) Channel {
+	// This method forces a deep-copy of the object to ensure immutability.
+	newChannel := channel.DeepCopy()
+	if patch.Name != nil {
+		newChannel.name = *patch.Name
+	}
+
+	// Question: How do we handle setting / deletion of Props?
+
+	return newChannel
+}
+
+// DeepCopy methods for model objects ensure immutability.
+func (channel *Channel) DeepCopy() Channel {
+	// Question: can we automate the implementation of these DeepCopy functions?
+	// Question: if not, can we use govet to ensure they are correct?
+
+	newChannel := Channel{
+		id: channel.id,
+		name: channel.name,
+	}
+
+	// For pointer types, copy the *value* not the pointer.
+	if channel.schemeID != nil {
+		*newChannel.schemeID = *channel.schemeID
+	}
+
+	// Question: how do we handle slices and maps on the model object, such as Props?
+
+	return newChannel
+}
+
+// Question: how do we create a brand new Channel model object?
+
+// Question: how do we populate a Channel model object when querying
+// the database in the store layer?
+
+// Question: do we provide any helper methods for JSON marshalling/unmarshalling?
```

Comment:
I think we going to need it, and if the properties are private, I think we can't do it externally using encoding/json library, so probably we need methods for marshal/unmarshall.

Reply:
One option for the json marshalling/unmarshalling seems to be this:

```go
struct channel {
  ID   string `json:"id"`
  Name string `json:"name"`
}

struct Channel {
  channel
}

func (c *Channel) UnmarshalJSON(b []byte) error {
    return json.Unmarshal(b, &d.channel)
}

func (c *Channel) Name() string {
  return c.channel.Name
}
```

Not sure how much I like that, but there it is for consideration anyway.

cc. @agnivade @jespino 

reference: https://stackoverflow.com/questions/11126793/json-and-dealing-with-unexported-fields

Reply:
Another option, because is code generation is to de-serialize to other object and then copy the fields. Or maybe deserialize to the model.ChannelInitializer struct, and use that struct to build the instance.

```go
struct ChannelInitializer {
  ID   string `json:"id"`
  Name string `json:"name"`
}

struct Channel {
  id       string
  name string
}

func ChannelFromJSON(b []byte) (*Channel, error) {
    var data ChannelInitializer
    err := json.Unmarshal(b, &data)
    if err != nil {
        return nil, err
    }
    return NewChannel(data), nil
}

func (c *Channel) Name() string {
  return c.name
}
```

Reply:
By the way, if we want to avoid performance implecations of ChannelInitializer here, we probably can directly use type casting instead of copying data as far as the whole data structure is identical.

Reply:
I like Jesus' idea. I think the performance implications of copying the fields is going to be absolutely negligible (we can easily benchmark it to get evidence). Here is my slight modifications on it:

```go
type ChannelInitializer struct {
  ID   string `json:"id"`
  Name string `json:"name"`
}

// Channel is auto-generated from ChannelInitializer.
type Channel struct{
  id   string
  name string
}

// UnmarshalJSON is auto-generated.
func (c *Channel) UnmarshalJSON(b []byte) (*Channel, error) {
    var data *ChannelInitializer
    err := json.Unmarshal(b, data)
    if err != nil {
        return nil, err
    }
    return NewChannel(data), nil
}

// NewChannel is auto-generated.
func NewChannel(input *ChannelInitializer) *Channel {
	return &Channel{
		...
	}
}

```

Overall, I like the idea of auto-generating our entire model package. This has several advantages like we can have an abstract spec for our model schema. And can specify field validation requirements in the spec itself.

For example, protobuf can specify required/optional fields as a struct tag. We can extend this idea to do something like that too. This way, we can even auto-generate the `IsValid()` methods.

Reply:
@agnivade I can see the `MarshalJSON` function working like this:

```
func (c *Channel) MarshalJSON() ([]byte, error) {
    return json.Marshall(&ChannelInitializer{
        ...
    })
}
```

But the signature for your `UnmarshalJSON()` function above isn't correct to implement the interface for the built-in `json` package to decode to. Can you see a way to make our "immutables" compatible with that, or would we just provide `ToJson()` / `FromJson()` functions and forget about being compatible with the golang interfaces for `json.Decode()/json.Encode()`?

Reply:
Woops .. my bad.

It should be 

```go
// UnmarshalJSON is auto-generated.
func (c *Channel) UnmarshalJSON(b []byte) (error) {
    var data *ChannelInitializer
    err := json.Unmarshal(b, data)
    if err != nil {
        return err
    }
    *c := *NewChannel(data)
     return nil
}
```

See: https://golang.org/pkg/encoding/json/#example__customMarshalJSON

Reply:
@agnivade about your UnmarshalJSON suggestion: how does this fit into 'immutable' model? We are basically allowing the user to overwrite all fields in the Channel object...